dapps-certification-signature-verification/src/IOHK/Certification/SignatureVerification.hs

```diff
@@ -0,0 +1,132 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+module IOHK.Certification.SignatureVerification
+  ( module CBOR
+  , verifyCIP30Signature
+  , Bech32Address(..)
+  , COSEKey
+  , COSESign1
+  , encodeHex
+  , decodeHex
+  ) where
+
+import IOHK.Certification.SignatureVerification.CBOR as CBOR
+  ( unfoldPayload
+  , unfoldPublicKey
+  , Address(..)
+  , Message(..)
+  , PublicKey(..)
+  , Signature(..)
+  , SignedMessage(..)
+  , UnfoldedPayload(..)
+  )
+import Cardano.Address
+  ( Address, unsafeMkAddress, fromBech32 )
+import Data.Text
+  ( Text )
+import Data.ByteString
+  ( ByteString, empty )
+import Crypto.Error
+  ( CryptoFailable(CryptoPassed, CryptoFailed) )
+import Control.Applicative
+  ((<|>))
+import Control.Monad
+  (unless)
+
+import qualified Data.ByteString.Base16 as Hexa
+import qualified Cardano.Address.Style.Shelley as Shelley
+import qualified Crypto.PubKey.Ed25519 as Ed25519
+import qualified Crypto.Hash.BLAKE2.BLAKE2b as Blake
+
+verifySignature :: PublicKey -> SignedMessage -> Signature -> Either String Bool
+verifySignature (PublicKey pubKey) (SignedMessage msg) (Signature sig) = do
+    key <- case Ed25519.publicKey pubKey of
+      CryptoFailed err -> Left $ "Failed to create Ed25519 public key: " <> show err
+      CryptoPassed res -> pure res
+    sig <- case Ed25519.signature sig of
+      CryptoFailed err -> Left $ "Failed to create Ed25519 signature: " <> show err
+      CryptoPassed res -> pure res
+    Right $ Ed25519.verify key msg sig
+
+type COSEKey = ByteString
+type COSESign1 = ByteString
+
+-- | Address and public key verification
+
+verifyCIP30Signature :: COSEKey -> COSESign1 -> Maybe Message -> Maybe Bech32Address -> Either String Bool
+verifyCIP30Signature coseKey coseSign1 mMsg mBech32Address = do
+  -- decode the CBOR-encoded signed payload
+  UnfoldedPayload{..} <- unfoldPayload coseSign1
+  -- decode the CBOR-encoded public key
+  pubKey <- unfoldPublicKey coseKey
+  -- verify the address
+  hashVerificationResp <- verifyHashes pubKey address mBech32Address
+  unless hashVerificationResp $ Left "Hash address verification failed"
+  -- verify the signature
+  verificationResp <- verifySignature pubKey signedMessage signature
+  Right $ verificationResp
+    -- if a message was provided, verify that it matches the message in the signed payload
+    && maybe True (== message) mMsg
+
+verifyHashes :: PublicKey -> CBOR.Address -> Maybe Bech32Address -> Either String Bool
```

Comment:
i'm thinking at two points: 
- aren't the resulted hashes in the hex format instead of Bech32?
- only the third argument seems to be in the Bech32 format which is also optional

Reply:
You're completely right.